plugin/pulpcore/plugin/stages/api.py

```diff
@@ -0,0 +1,73 @@
+import asyncio
+from gettext import gettext as _
+
+
+class BaseStage:
+    """
+    The base class for all Stages API stages.
+
+    To make a stage, inherit from this class and implement :meth:`__call__` on the subclass.
+    """
+
+    async def __call__(self, in_q, out_q):
+        """
+        The coroutine that is run as part of this stage.
+
+        Args:
+            in_q (:class:`asyncio.Queue`): The queue to receive items from the previous stage.
+            out_q (:class:`asyncio.Queue`): The queue to put handled items into for the next stage.
+
+        Returns:
+            The coroutine that runs this stage.
+
+        """
+        raise NotImplementedError(_('A plugin writer must implement this method'))
+
+
+async def create_pipeline(stages, maxsize=100):
+    """
+    Creates a Stages API linear pipeline from the list `stages` and return as a single coroutine.
+
+    Each stage is a coroutine and reads from an input :class:`asyncio.Queue` and writes to an output
+    :class:`asyncio.Queue`. When the stage is ready to shutdown it writes a `None` to the output
+    queue. Here is an example of the simplest stage that only passes data.
+
+    >>> async def my_stage(in_q, out_q):
+    >>>     while True:
+    >>>         item = await in_q.get()
+    >>>         if item is None:  # Check if the previous stage is shutdown
+    >>>             break
+    >>>         await out_q.put(item)
+    >>>     await out_q.put(None)  # this stage is shutdown so send 'None'
+
+    Args:
+        stages (list of coroutines): A list of Stages API compatible coroutines.
+        maxsize (int): The maximum amount of items a queue between two stages should hold. Optional
+            and defaults to 100.
+
+    Returns:
+        A single coroutine that can be used to run, wait, or cancel the entire pipeline with.
+    """
+    in_q = None
+    futures = []
+    for stage in stages:
+        out_q = asyncio.Queue(maxsize=maxsize)
+        futures.append(stage(in_q, out_q))
+        in_q = out_q
+    await asyncio.gather(*futures)
+
+
+class EndStage(BaseStage):
```

Comment:
Big :+1:  for making the stages objects!  This will unlock a lot of potential.

At the risk of being pedantic :(

Why not just `Stage`?

The rare cases I see the *Base* prefix, it strikes me as odd. The class name is a *classification* of objects and should read naturally using the **IsA** lexicon.

For example, the natural way to describe an `QueryExistingArtifacts` is to say:

 *"A QueryExistingArtifacts` IsA Stage"*.

not

 *"A QueryExistingArtifacts` IsA BaseStage"*.

I don't think designating this as a *base* class though naming adds value.  This isn't a blocker but, IMHO, it would be more correct in the classic object modeling sense.

Reply:
I think s/BaseStage/Stage/ is a great change. I'll make it on my next push along w/ the last remaining fixes.